compiler/rustc_resolve/src/lib.rs

```diff
@@ -1584,6 +1586,57 @@ impl<'a> Resolver<'a> {
         });
     }
 
+    fn check_reexport_ambiguities(&mut self) {
+        for module in self.arenas.local_modules().iter() {
+            if let Some(_) = module.opt_def_id() {
+                let mut ambiguous_reexports = FxHashMap::default();
+
+                module.for_each_child(self, |this, ident, ns, binding| {
+                    if let NameBindingKind::Import { import: reexport, .. } = binding.kind
+                    && let res = binding.res().expect_non_local::<Res>()
+                    && res != def::Res::Err {
+                        if let Some((ambiguous_binding, _)) = binding.ambiguity
+                        && let NameBindingKind::Import { import: ambiguous_reexport, .. } = ambiguous_binding.kind {
+                            if let Some(did) = binding.res().opt_def_id()
+                            && let Some(local_did) = did.as_local()
+                            && let Some(amb_did) = ambiguous_binding.res().opt_def_id()
+                            && let Some(amb_local_did) = amb_did.as_local()
+                            {
+                                debug!("check_reexport_ambiguities: did = {:?}, amb_did = {:?}", did, amb_did);
+                                debug!("check_reexport_ambiguities: did.is_exported = {:?}, amb_did.is_exported = {:?}", this.effective_visibilities.is_exported(local_did), this.effective_visibilities.is_exported(amb_local_did));
+                                if this.effective_visibilities.is_exported(local_did)
+                                && this.effective_visibilities.is_exported(amb_local_did) {
```

Comment:
You are checking `is_exported` for wrong IDs here, ID of the import itself should be used instead of ID of the item that is ultimately imported.

Reply:
By ID of the import itself you mean `binding.res().def_id().as_local()`?

Reply:
No, `binding.res()` is a resolution of the import's destination.
I mean this thing - `NameBindingKind::Import { import, .. }` -> `import.id()` converted to `LocalDefId`, that's the ID of the import itself.

Reply:
Ah, I will try to figure out how I can get a LocalDefId from that, thanks

Reply:
I think it's because the LocalDefId corresponding to the re-export doesn't exist in the Effective Visibilities table.

```rust
DEBUG rustc_resolve reexport_def_id: DefId(0:7)
DEBUG rustc_resolve reexport_def_id: this.effective_visibilities.is_exported(reexport_def_id) = false
```

but the effective visibility table has no entry for `DefId(0:7)`:

```rust
 INFO rustc_resolve::effective_visibilities resolve::effective_visibilities: EffectiveVisibilities {
    map: {
        DefId(0:0): EffectiveVisibility {
            direct: Public,
            reexported: Public,
            reachable: Public,
            reachable_through_impl_trait: Public,
        },
        DefId(0:5): EffectiveVisibility {
            direct: Public,
            reexported: Public,
            reachable: Public,
            reachable_through_impl_trait: Public,
        },
        DefId(0:2): EffectiveVisibility {
            direct: Restricted(
                DefId(0:0),
            ),
            reexported: Restricted(
                DefId(0:0),
            ),
            reachable: Restricted(
                DefId(0:0),
            ),
            reachable_through_impl_trait: Restricted(
                DefId(0:0),
            ),
        },
        DefId(0:4): EffectiveVisibility {
            direct: Public,
            reexported: Public,
            reachable: Public,
            reachable_through_impl_trait: Public,
        },
        DefId(0:9): EffectiveVisibility {
            direct: Public,
            reexported: Public,
            reachable: Public,
            reachable_through_impl_trait: Public,
        },
        DefId(0:6): EffectiveVisibility {
            direct: Public,
            reexported: Public,
            reachable: Public,
            reachable_through_impl_trait: Public,
        },
        DefId(0:3): EffectiveVisibility {
            direct: Public,
            reexported: Public,
            reachable: Public,
            reachable_through_impl_trait: Public,
        },
    },
}
```

Reply:
@jieyouxu 
Sigh, it's due to this condition - https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/effective_visibilities.rs#L126

We know that the ambiguity reexport won't actually reexport anything, so we don't mark it as exported.
If you change
```rust
pub use foo::*;
pub use bar::*;
```
to export at least something (not only ambiguous `X`) then it should appear in the table.

Reply:
It should be ok to remove the ambiguity condition from https://github.com/rust-lang/rust/blob/master/compiler/rustc_resolve/src/effective_visibilities.rs#L126, the effective visibility table is a conservative estimate anyway, and can contain additional entries.

Reply:
Single test failure being the auxillary build for 
`tests/ui/imports/glob-conflict-cross-crate.rs`, `tests/ui/imports/auxiliary/glob-conflict.rs`

The auxillary file contains

```rust
mod m1 {
    pub fn f() {}
}
mod m2 {
    pub fn f(_: u8) {}
}

pub use m1::*;
pub use m2::*;

pub mod glob {
    pub use *;
}
```

The main test file contains:

```rust
// aux-build:glob-conflict.rs

extern crate glob_conflict;

fn main() {
    glob_conflict::f(); //~ ERROR cannot find function `f` in crate `glob_conflict`
    glob_conflict::glob::f(); //~ ERROR cannot find function `f` in module `glob_conflict::glob`
}
```

which I think this lint correctly catches. Must the auxillary file successfully compile? Since this test is testing that the ambiguous glob re-export isn't actually being actually exported across crate boundaries. But the lint is catching the ambiguous re-export before it can successfully compile...

Reply:
>I guess I could explicitly allow ambiguous re-exports for this particular auxillary file?

Yes

Reply:
While the UI tests pass, I'm still getting errors trying to compile `diesel`. Will reinvestigate why the lint is fired on the re-exports.